File: bindings/nodejs/src/keys.ts
/**
 * HSM Key Storage operations
 * @module keys
 */

import { getLib, createSizeT, readSizeT } from './ffi';
import { checkError, QuacError, ErrorCode } from './errors';
import { KeyType, KeyUsage, KeyInfo } from './types';
import type { Device } from './device';

/**
 * HSM Key Storage interface
 *
 * Provides secure key storage operations on the hardware.
 *
 * @example
 * ```typescript
 * const keys = device.keys;
 *
 * // Store a key
 * const keyData = crypto.randomBytes(32);
 * keys.store(0, KeyType.Secret, 0, KeyUsage.Encrypt | KeyUsage.Decrypt, 'my-key', keyData);
 *
 * // Load a key
 * const loaded = keys.load(0);
 *
 * // Get key info
 * const info = keys.getInfo(0);
 * console.log(`Key: ${info.label}`);
 *
 * // Delete a key
 * keys.delete(0);
 * ```
 */
export class Keys {
    private readonly device: Device;

    constructor(device: Device) {
        this.device = device;
    }

    /**
     * Store a key in the HSM
     * @param slot - Slot number
     * @param keyType - Type of key
     * @param algorithm - Algorithm identifier
     * @param usage - Allowed key usages
     * @param label - Human-readable label
     * @param keyData - Key data
     */
    store(
        slot: number,
        keyType: KeyType,
        algorithm: number,
        usage: KeyUsage,
        label: string,
        keyData: Buffer
    ): void {
        const lib = getLib();
        const handle = this.device.getHandle();

        const result = lib.quac_key_store(
            handle,
            slot,
            keyType,
            algorithm,
            usage,
            label,
            keyData,
            keyData.length
        );
        checkError(result);
    }

    /**
     * Load a key from the HSM
     * @param slot - Slot number
     * @returns Key data
     */
    load(slot: number): Buffer {
        const lib = getLib();
        const handle = this.device.getHandle();

        // Allocate max possible key size
        const keyData = Buffer.alloc(8192);
        const keyDataLen = createSizeT(keyData.length);

        const result = lib.quac_key_load(handle, slot, keyData, keyDataLen);
        checkError(result);

        return keyData.subarray(0, readSizeT(keyDataLen));
    }

    /**
     * Get information about a stored key
     * @param slot - Slot number
     * @returns Key information
     */
    getInfo(slot: number): KeyInfo {
        const lib = getLib();
        const handle = this.device.getHandle();

        const infoBuffer = Buffer.alloc(256);
        const result = lib.quac_key_get_info(handle, slot, infoBuffer);
        checkError(result);

        return {
            slot: infoBuffer.readUInt32LE(0),
            keyType: infoBuffer.readInt32LE(4) as KeyType,
            algorithm: infoBuffer.readInt32LE(8),
            usage: infoBuffer.readUInt32LE(12) as KeyUsage,
            label: infoBuffer.toString('utf8', 16, 80).replace(/\0/g, '').trim(),
            size: infoBuffer.readUInt32LE(80),
            createdAt: new Date(infoBuffer.readUInt32LE(84) * 1000),
        };
    }

    /**
     * Delete a key from the HSM
     * @param slot - Slot number
     */
    delete(slot: number): void {
        const lib = getLib();
        const handle = this.device.getHandle();

        const result = lib.quac_key_delete(handle, slot);
        checkError(result);
    }

    /**
     * List all occupied slots
     * @returns Array of slot numbers
     */
    list(): number[] {
        const lib = getLib();
        const handle = this.device.getHandle();

        // Get slot count first
        const slotCount = this.getSlotCount();
        const slots = Buffer.alloc(slotCount * 4);
        const count = createSizeT(slotCount);

        const result = lib.quac_key_list(handle, slots, count);
        checkError(result);

        const actualCount = readSizeT(count);
        const slotNumbers: number[] = [];
        for (let i = 0; i < actualCount; i++) {
            slotNumbers.push(slots.readUInt32LE(i * 4));
        }

        return slotNumbers;
    }

    /**
     * Get total number of key slots
     */
    getSlotCount(): number {
        const lib = getLib();
        const handle = this.device.getHandle();

        const count = lib.quac_key_get_slot_count(handle);
        if (count < 0) {
            throw QuacError.fromCode(count);
        }
        return count;
    }

    /**
     * Get the first free slot
     * @returns Free slot number, or -1 if no slots available
     */
    getFreeSlot(): number {
        const lib = getLib();
        const handle = this.device.getHandle();

        const slot = lib.quac_key_get_free_slot(handle);
        return slot;
    }

    /**
     * Check if a slot is occupied
     * @param slot - Slot number
     */
    isSlotOccupied(slot: number): boolean {
        try {
            this.getInfo(slot);
            return true;
        } catch (e) {
            if (e instanceof QuacError && e.code === ErrorCode.SlotEmpty) {
                return false;
            }
            throw e;
        }
    }

    /**
     * Clear all keys from the HSM
     * @warning This permanently deletes all stored keys!
     */
    clearAll(): void {
        const lib = getLib();
        const handle = this.device.getHandle();

        const result = lib.quac_key_clear_all(handle);
        checkError(result);
    }

    /**
     * Store a KEM key pair
     * @param publicKeySlot - Slot for public key
     * @param secretKeySlot - Slot for secret key
     * @param algorithm - Algorithm identifier
     * @param label - Label prefix
     * @param publicKey - Public key data
     * @param secretKey - Secret key data
     */
    storeKemKeypair(
        publicKeySlot: number,
        secretKeySlot: number,
        algorithm: number,
        label: string,
        publicKey: Buffer,
        secretKey: Buffer
    ): void {
        this.store(
            publicKeySlot,
            KeyType.Public,
            algorithm,
            KeyUsage.Encrypt,
            `${label}-pub`,
            publicKey
        );
        this.store(
            secretKeySlot,
            KeyType.Private,
            algorithm,
            KeyUsage.Decrypt,
            `${label}-sec`,
            secretKey
        );
    }

    /**
     * Store a signature key pair
     * @param publicKeySlot - Slot for public key
     * @param secretKeySlot - Slot for secret key
     * @param algorithm - Algorithm identifier
     * @param label - Label prefix
     * @param publicKey - Public key data
     * @param secretKey - Secret key data
     */
    storeSignKeypair(
        publicKeySlot: number,
        secretKeySlot: number,
        algorithm: number,
        label: string,
        publicKey: Buffer,
        secretKey: Buffer
    ): void {
        this.store(
            publicKeySlot,
            KeyType.Public,
            algorithm,
            KeyUsage.Verify,
            `${label}-pub`,
            publicKey
        );
        this.store(
            secretKeySlot,
            KeyType.Private,
            algorithm,
            KeyUsage.Sign,
            `${label}-sec`,
            secretKey
        );
    }
}

File: bindings/rust/src/keys.rs
//! HSM Key Storage operations.
//!
//! This module provides secure key storage in the hardware HSM.

use crate::device::Device;
use crate::error::{check_error, Result, QuacError, ErrorCode};
use crate::ffi;
use crate::types::{KeyType, KeyUsage};

use std::ffi::CString;

/// Information about a stored key.
#[derive(Debug, Clone)]
pub struct KeyInfo {
    /// Slot number
    pub slot: u32,
    /// Key type
    pub key_type: KeyType,
    /// Algorithm identifier
    pub algorithm: i32,
    /// Usage flags
    pub usage: KeyUsage,
    /// Key label
    pub label: String,
    /// Creation timestamp (Unix epoch)
    pub created: u64,
    /// Whether the key can be exported
    pub exportable: bool,
}

impl KeyInfo {
    pub(crate) fn from_ffi(info: &ffi::quac_key_info_t) -> Self {
        use std::ffi::CStr;
        
        Self {
            slot: info.slot,
            key_type: KeyType::from_raw(info.key_type).unwrap_or(KeyType::Secret),
            algorithm: info.algorithm,
            usage: KeyUsage::from_bits_truncate(info.usage),
            label: unsafe {
                CStr::from_ptr(info.label.as_ptr())
                    .to_string_lossy()
                    .into_owned()
            },
            created: info.created,
            exportable: info.exportable != 0,
        }
    }
}

/// Key storage (HSM) subsystem.
///
/// Provides secure key storage in the hardware security module.
///
/// # Example
///
/// ```no_run
/// use quantacore::{initialize, open_first_device, KeyType, KeyUsage};
///
/// initialize().unwrap();
/// let device = open_first_device().unwrap();
/// let keys = device.keys();
///
/// // Store a key
/// let key_data = vec![0u8; 32]; // Your key material
/// keys.store(
///     0,  // slot
///     KeyType::Secret,
///     0,  // algorithm
///     KeyUsage::ENCRYPT | KeyUsage::DECRYPT,
///     "my-key",
///     &key_data,
/// ).unwrap();
///
/// // Load the key
/// let loaded = keys.load(0).unwrap();
///
/// // Get key info
/// let info = keys.get_info(0).unwrap();
/// println!("Key label: {}", info.label);
///
/// // Delete the key
/// keys.delete(0).unwrap();
/// ```
#[derive(Clone)]
pub struct Keys {
    device: Device,
}

impl Keys {
    /// Create a new keys subsystem handle.
    pub(crate) fn new(device: Device) -> Self {
        Self { device }
    }

    /// Store a key in the HSM.
    ///
    /// # Arguments
    ///
    /// * `slot` - The slot number to store the key in
    /// * `key_type` - The type of key
    /// * `algorithm` - The algorithm identifier
    /// * `usage` - Allowed usage flags
    /// * `label` - A human-readable label (max 63 chars)
    /// * `key_data` - The raw key material
    pub fn store(
        &self,
        slot: u32,
        key_type: KeyType,
        algorithm: i32,
        usage: KeyUsage,
        label: &str,
        key_data: &[u8],
    ) -> Result<()> {
        let c_label = CString::new(label).map_err(|_| {
            QuacError::InvalidParameter("Label contains null bytes".into())
        })?;

        let result = unsafe {
            ffi::quac_key_store(
                self.device.handle(),
                slot,
                key_type.to_raw(),
                algorithm,
                usage.bits(),
                c_label.as_ptr(),
                key_data.as_ptr(),
                key_data.len(),
            )
        };

        check_error(result)
    }

    /// Load a key from the HSM.
    ///
    /// # Arguments
    ///
    /// * `slot` - The slot number to load from
    ///
    /// # Returns
    ///
    /// The raw key material.
    pub fn load(&self, slot: u32) -> Result<Vec<u8>> {
        // First get info to determine key size
        let _info = self.get_info(slot)?;
        let max_size = 8192; // Maximum reasonable key size

        let mut key_data = vec![0u8; max_size];
        let mut key_len = max_size;

        let result = unsafe {
            ffi::quac_key_load(
                self.device.handle(),
                slot,
                key_data.as_mut_ptr(),
                &mut key_len,
            )
        };

        check_error(result)?;
        key_data.truncate(key_len);

        Ok(key_data)
    }

    /// Get information about a stored key.
    ///
    /// # Arguments
    ///
    /// * `slot` - The slot number
    ///
    /// # Returns
    ///
    /// Information about the key in that slot.
    pub fn get_info(&self, slot: u32) -> Result<KeyInfo> {
        let mut info = ffi::quac_key_info_t::default();
        let result = unsafe {
            ffi::quac_key_get_info(self.device.handle(), slot, &mut info)
        };
        check_error(result)?;
        Ok(KeyInfo::from_ffi(&info))
    }

    /// Delete a key from the HSM.
    ///
    /// # Arguments
    ///
    /// * `slot` - The slot number to delete
    pub fn delete(&self, slot: u32) -> Result<()> {
        let result = unsafe { ffi::quac_key_delete(self.device.handle(), slot) };
        check_error(result)
    }

    /// List all occupied key slots.
    ///
    /// # Returns
    ///
    /// A vector of slot numbers that contain keys.
    pub fn list(&self) -> Result<Vec<u32>> {
        let capacity = self.get_slot_count()? as usize;
        let mut slots = vec![0u32; capacity];
        let mut count = capacity;

        let result = unsafe {
            ffi::quac_key_list(
                self.device.handle(),
                slots.as_mut_ptr(),
                &mut count,
            )
        };

        check_error(result)?;
        slots.truncate(count);

        Ok(slots)
    }

    /// Get the total number of key slots.
    pub fn get_slot_count(&self) -> Result<u32> {
        let result = unsafe { ffi::quac_key_get_slot_count(self.device.handle()) };
        if result < 0 {
            check_error(result)?;
        }
        Ok(result as u32)
    }

    /// Find the first free slot.
    ///
    /// # Returns
    ///
    /// The slot number, or an error if no free slots.
    pub fn get_free_slot(&self) -> Result<u32> {
        let result = unsafe { ffi::quac_key_get_free_slot(self.device.handle()) };
        if result < 0 {
            return Err(QuacError::from_code(ErrorCode::KeyNotFound));
        }
        Ok(result as u32)
    }

    /// Clear all keys from the HSM.
    ///
    /// **WARNING**: This permanently deletes all stored keys!
    pub fn clear_all(&self) -> Result<()> {
        let result = unsafe { ffi::quac_key_clear_all(self.device.handle()) };
        check_error(result)
    }

    /// Check if a slot is occupied.
    pub fn is_slot_occupied(&self, slot: u32) -> Result<bool> {
        match self.get_info(slot) {
            Ok(_) => Ok(true),
            Err(QuacError::Native { code: ErrorCode::KeyNotFound, .. }) => Ok(false),
            Err(e) => Err(e),
        }
    }

    /// Store a key pair (public + secret).
    ///
    /// This is a convenience method that stores both keys in adjacent slots.
    ///
    /// # Arguments
    ///
    /// * `slot` - The base slot (public key goes here, secret at slot+1)
    /// * `algorithm` - Algorithm identifier
    /// * `label` - Label prefix
    /// * `public_key` - Public key bytes
    /// * `secret_key` - Secret key bytes
    pub fn store_keypair(
        &self,
        slot: u32,
        algorithm: i32,
        label: &str,
        public_key: &[u8],
        secret_key: &[u8],
    ) -> Result<()> {
        self.store(
            slot,
            KeyType::Public,
            algorithm,
            KeyUsage::VERIFY,
            &format!("{}-pub", label),
            public_key,
        )?;

        self.store(
            slot + 1,
            KeyType::Private,
            algorithm,
            KeyUsage::SIGN,
            &format!("{}-priv", label),
            secret_key,
        )?;

        Ok(())
    }
}

impl std::fmt::Debug for Keys {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Keys").finish()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_key_usage_flags() {
        let usage = KeyUsage::ENCRYPT | KeyUsage::DECRYPT;
        assert!(usage.contains(KeyUsage::ENCRYPT));
        assert!(usage.contains(KeyUsage::DECRYPT));
        assert!(!usage.contains(KeyUsage::SIGN));
    }
}
